File: uri/ad-hock/jollo/1321.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    while (true) {
        int a, b, c, x, y;
        if (!(cin >> a >> b >> c >> x >> y)) {
            break;
        }
        if (a == 0 && b == 0 && c == 0 && x == 0 && y == 0) {
            break;
        }
        vector<int> princess = {a, b, c};
        vector<int> prince = {x, y};
        vector<bool> used(53, false);
        for (int v : princess) {
            used[v] = true;
        }
        for (int v : prince) {
            used[v] = true;
        }
        sort(princess.begin(), princess.end());
        int answer = -1;
        for (int candidate = 1; candidate <= 52; ++candidate) {
            if (used[candidate]) {
                continue;
            }
            vector<int> order = prince;
            order.push_back(candidate);
            sort(order.begin(), order.end());
            int minWins = 3;
            do {
                vector<int> available = princess;
                int princeWins = 0;
                for (int card : order) {
                    auto it = upper_bound(available.begin(), available.end(), card);
                    if (it == available.end()) {
                        available.erase(available.begin());
                        ++princeWins;
                    } else {
                        available.erase(it);
                    }
                }
                minWins = min(minWins, princeWins);
            } while (next_permutation(order.begin(), order.end()));
            if (minWins >= 2) {
                answer = candidate;
                break;
            }
        }
        cout << answer << "\n";
    }

    return 0;
}

File: uri/ad-hock/jollo/1321.rs
use std::io::{self, Read};

fn next_permutation(data: &mut [i32]) -> bool {
    if data.len() < 2 {
        return false;
    }
    let mut i = data.len() - 1;
    while i > 0 && data[i - 1] >= data[i] {
        i -= 1;
    }
    if i == 0 {
        return false;
    }
    let mut j = data.len() - 1;
    while data[j] <= data[i - 1] {
        j -= 1;
    }
    data.swap(i - 1, j);
    data[i..].reverse();
    true
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut iter = input.split_whitespace();
    loop {
        let a = match iter.next() {
            Some(v) => v.parse::<i32>().unwrap(),
            None => break,
        };
        let b = iter.next().unwrap().parse::<i32>().unwrap();
        let c = iter.next().unwrap().parse::<i32>().unwrap();
        let x = iter.next().unwrap().parse::<i32>().unwrap();
        let y = iter.next().unwrap().parse::<i32>().unwrap();
        if a == 0 && b == 0 && c == 0 && x == 0 && y == 0 {
            break;
        }
        let mut princess = vec![a, b, c];
        princess.sort();
        let prince = vec![x, y];
        let mut used = [false; 53];
        for v in princess.iter() {
            used[*v as usize] = true;
        }
        for v in prince.iter() {
            used[*v as usize] = true;
        }
        let mut answer = -1;
        for candidate in 1..=52 {
            if used[candidate as usize] {
                continue;
            }
            let mut order = vec![prince[0], prince[1], candidate as i32];
            order.sort();
            let mut min_wins = 3;
            loop {
                let mut available = princess.clone();
                let mut prince_wins = 0;
                for &card in order.iter() {
                    let mut idx = None;
                    for (pos, &pv) in available.iter().enumerate() {
                        if pv > card {
                            idx = Some(pos);
                            break;
                        }
                    }
                    if let Some(pos) = idx {
                        available.remove(pos);
                    } else {
                        available.remove(0);
                        prince_wins += 1;
                    }
                }
                if prince_wins < min_wins {
                    min_wins = prince_wins;
                }
                if !next_permutation(&mut order) {
                    break;
                }
            }
            if min_wins >= 2 {
                answer = candidate as i32;
                break;
            }
        }
        println!("{}", answer);
    }
}
